### PR TITLE
feat: Implement fallback strategy in tier balancer #588

### PR DESCRIPTION
**Pull request by astsiapanay**

### CODE REVIEW

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java [hunk 1]
```diff
@@ -28,30 +31,32 @@
...
 
-    public UpstreamRoute(LoadBalancer<UpstreamState> balancer, int maxRetryAttempts) {
+    private final Predicate<Upstream> fallbackPredicate;
+
+    public UpstreamRoute(TieredBalancer balancer, int maxRetryAttempts) {
         this.balancer = balancer;
         this.maxRetryAttempts = maxRetryAttempts;
-        this.upstreamState = balancer.next();
-        this.attemptCount = upstreamState == null ? 0 : 1;
+        Set<Upstream> usedUpstreams = new HashSet<>();
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java]: I do not get it. I expected that you would update it when next() is called. But instead, it is updated in predicate.

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java [hunk 2]
```diff
@@ -94,31 +99,29 @@ public Upstream get() {
...
+        long retryAfter = retrieveRetryAfterSeconds(response);
+        HttpStatus status = HttpStatus.fromStatusCode(response.statusCode());
+        fail(status, retryAfter);
     }
 
     public void succeed() {
-        if (upstreamState != null) {
-            upstreamState.succeeded();
-        }
+        balancer.succeed(upstream);
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java:117]: please make sure that API is used as expected with Objects.requireNonNull

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java [hunk 3]
```diff
@@ -94,31 +99,29 @@ public Upstream get() {
      * @param status - response http status; typically, 5xx or 429
      * @param retryAfterSeconds - the amount of seconds after which upstream should be available; if status 5xx this value ignored
      */
-    public void fail(HttpStatus status, long retryAfterSeconds) {
-        if (upstreamState != null) {
-            upstreamState.failed(status, retryAfterSeconds);
-        }
+    void fail(HttpStatus status, long retryAfterSeconds) {
+        balancer.fail(upstream, status, retryAfterSeconds);
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java]: Make sure that API is used as expected with Objects.requireNonNull(upstream)

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java [hunk 4]
```diff
@@ -98,16 +98,34 @@ private String getKey(Deployment deployment) {
...
+        long lastAccessTime;
+
+        /**
+         * Note. The value is taken from {@link Deployment#getMaxRetryAttempts()} or {@link Route#getMaxRetryAttempts()}
+         */
+        final int maxRetryAttempts;
+
+        final List<Upstream> upstreams;
+
+        public BalancerWrapper(TieredBalancer balancer, int maxRetryAttempts, List<Upstream> upstreams) {
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java]: You can simplify by taking just upstreams and maxRetryAttemps and create a balancer here.

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamState.java [hunk 5]
```diff
@@ -38,30 +46,45 @@ public UpstreamState(Upstream upstream, int errorsThreshold) {
...
-        if (status.is5xx()) {
-            if (++errorCount >= errorsThreshold) {
-                retryAfter = System.currentTimeMillis()
-                             + Math.min(INITIAL_BACKOFF_DELAY_MS * (1L << errorCount), MAX_BACKOFF_DELAY_MS);
+        } else if (status.is5xx()) {
+            if (source == RetryAfterSource.CORE) {
+                errorCount++;
+                if (errorCount >= errorsThreshold) {
+                    // exponential or max backoff in seconds
+                    retryAfterSeconds = Math.min(1L << errorCount, MAX_BACKOFF_DELAY_SEC);
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamState.java]: Could give a negative, e.g. (1 << 31): -2147483648

Comment (astsiapanay) [reply server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamState.java]: I think we can remove `errorCount` and set retryAfter immediately as we do 429

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java (hunk 4 above)
Comment (astsiapanay) [reply server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java (hunk 2 above)
Comment (astsiapanay) [reply server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java:117]: fixed

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java (hunk 3 above)
Comment (astsiapanay) [reply server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java (hunk 1 above)
Comment (astsiapanay) [reply server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamState.java (hunk 5 above)
Comment (astsiapanay) [reply server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamState.java]: fixed

**Review by artsiomkorzun — APPROVED** [review]

**Comment by astsiapanay** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-593.nightly-test.deltixhub.io
>E2E tests status: success

**Pull request merged**